File: bubbleSort.cpp
#include <iostream>
#include <vector>
using namespace std;

class BubbleSort {
public:
    // Function to perform bubble sort
    // Takes a vector of integers as input
    void sort(vector<int>& arr) {
        int n = arr.size();  // Get the size of the array

        // Outer loop to traverse all elements of the array
        for (int i = 0; i < n-1; i++) {
            // Inner loop to compare each pair of adjacent elements
            for (int j = 0; j < n-i-1; j++) {
                // If the current element is greater than the next element, swap them
                if (arr[j] > arr[j+1]) {
                    swap(arr[j], arr[j+1]);  // Swap the elements
                }
            }
        }
    }

    // Function to print the array
    void printArray(const vector<int>& arr) {
        for (int num : arr) {
            cout << num << " ";  // Print each element
        }
        cout << endl;
    }
};

int main() {
    vector<int> arr = {64, 25, 12, 22, 11};  // Sample array
    BubbleSort sorter;  // Create an object of BubbleSort class

    cout << "Original array: ";
    sorter.printArray(arr);  // Print original array

    sorter.sort(arr);  // Call the sort function to sort the array

    cout << "Sorted array: ";
    sorter.printArray(arr);  // Print the sorted array

    return 0;
}

File: learn/sorting/bubble-sort/bubble_sort.rs
pub struct BubbleSort;

impl BubbleSort {
    pub fn sort(arr: &mut [i32]) {
        let mut n = arr.len();
        let mut swapped = true;
        while swapped {
            swapped = false;
            for i in 1..n {
                if arr[i - 1] > arr[i] {
                    arr.swap(i - 1, i);
                    swapped = true;
                }
            }
            n = n.saturating_sub(1);
        }
    }
}
